Refactor of pizza crate code and various other small pizza related code changes

flowercuco opened the pull request:
## About The Pull Request

The first and biggest thing that is in this pull request is changing the old pizza crate code from something that is hard to parse and full of blahblach = C stuff into things that I believe are more in line with modern standards, which adds the potential for expansion or additions to pizza crates with more ease if, for example, someone wanted to make a flatbread crate or something. 

I've also changed all of the lizard flatbreads from just /food/pizza/ into being /food/pizza/flatbread so that I could fax them, along with sliced bread. 

Then I went in on pizzabox code adding various screentips and balloon alerts so that holding stacks of pizza and manipulating them doesnt fill your chat with stuff and finally to pizza code to add screentips.
![dreamseeker_SF20VOXlNJ](https://user-images.githubusercontent.com/116288367/201151348-379393da-e807-45fd-9643-a2582ecd27a3.png)
![dreamseeker_wtROqx0gEc](https://user-images.githubusercontent.com/116288367/201151349-b84e031f-c94f-4265-962b-9151fa573b1e.png)
![dreamseeker_2dfG6QkV4R](https://user-images.githubusercontent.com/116288367/201155518-aba1ccf2-1fa2-4289-b766-ddae15e8bf1c.png)

EDIT: I've changed the code so that instead of using the tool context screentips add element, it adds screentips via the processable add element which means all processable food now has a default screentip of "process" as a result of this PR

## Why It's Good For The Game

Old pizza crate code was unclear and now is clear so it should be easier to know what to mess with if someone wants to change them further or add or expand them. 
We all like screentips!
Old pizzaboxes had way too many unnecessary messages getting printed to chat.
You can fax root flatbread but not flatbread with toppings on them and while that can make a kind of sense you can fax pizza slices with toppings on them! Bread slices just also makes sense to add while I'm at it.

## Changelog

:cl:
add: you can fax flatbreads and slices of bread now, changed all flatbreads into children of pizza/flatbread to allow for this
qol: adds screentips to various pizzabox action (stacking, when you can take pizzas out, pizza box bomb stuff)
qol: adds a way to add screentips directly into processable food via the add element processable proc
refactor: pizza crates work the same but more clearly and more variable
/:cl:

flowercuco committed:
adds flatbreads and breadslices to hacked fax whitelist
makes all flatbreads a subtype of pizza other than root flatbread which is a subtybe of bread
refactors pizza crate's funny properties to be more clear and slightly more adaptable

flowercuco committed:
changes defines to variables for use in any other potential pizza crates

flowercuco committed:
deconstructs pizzaboxes with right mouse click
adds some screentips

flowercuco committed:
finished screentips

flowercuco committed:
adds screentips to various pizza box activities
adds screentips to cutting pizza that can be cut into slices and flattening pizzaslices
changes a bunch of pizza bot stuff that doesnt need to be in the chat into balloon alerts or just removes them completely
you turn an empty pizza box into cardboard with RMB now

flowercuco committed:
removes _flatbread from lizard flatbreads that are now under the /flatbread/ header and changes recipies to match

dalmationer commented:
A sol's crate!?

flowercuco committed:
replaces instances of boxes.len to length(boxes)  across the entire file
removes unneccesary brackets
changes syntax on a ballons alert

flowercuco commented on code/game/objects/items/food/pizza.dm:
```diff
@@ -390,3 +404,12 @@
 	foodtypes = TOXIC
+
+///screentips for pizza
+/obj/item/food/pizza/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+
+	if(held_item?.tool_behaviour == TOOL_KNIFE || TOOL_SCALPEL || TOOL_SAW)
+		if(slice_type == null)
+			return
+		context[SCREENTIP_CONTEXT_LMB] = "Slice pizza"
+		return CONTEXTUAL_SCREENTIP_SET
```
i will look into the feasibility of (me personally) trying to add contextual screentips to the element because your solution wrt tools is what i was hoping i would not have to do haha. 

thank you!

flowercuco replied:
it turns out that someone WAS me! and only with medium help from @Jacquerel !!!

flowercuco committed:
adds context screentips to all processable food with the default verb process and the ability to change the verb into whatever actually matches

san7890 commented on code/datums/elements/food/processable.dm:
```diff
@@ -54,3 +62,22 @@
 		examine_list += span_notice("It can be turned into \a [initial(result_atom_type.name)] with <b>[tool_behaviour_name(tool_behaviour)]</b>!")
+
+///Adds context sensitivy directly to the processable file for screentips
+/datum/element/processable/proc/on_requesting_context_from_item(
+	datum/source,
+	list/context,
+	obj/item/held_item,
+	mob/user,
+)
```
```suggestion
/datum/element/processable/proc/on_requesting_context_from_item(datum/source, list/context, obj/item/held_item, mob/user)
```

you don't have THAT many arguments. also, might be worthwhile to DMdoc what each of these mean/do

flowercuco replied:
let me know if my documentation is off or wrong, but i think that theyre correct

san7890 replied:
The format looks alright to my eyes

san7890 commented on code/modules/cargo/packs/organic.dm:
```diff
@@ -185,52 +181,58 @@
+	anomalous_box_provided = TRUE
+	log_game("An anomalous pizza box was provided in a pizza crate at during cargo delivery.")
+	if(prob(50))
+		addtimer(CALLBACK(src, .proc/anomalous_pizza_report), rand(300, 1800))
```
```suggestion
		addtimer(CALLBACK(src, .proc/anomalous_pizza_report), rand(30 SECONDS, 180 SECONDS))
```
labelling units helps people know the exact intent, may as well touch it up while you're doing stuff ehre

san7890 commented on code/datums/elements/food/processable.dm:
```diff
@@ -12,8 +12,11 @@
+	var/screentip_verb
 
-/datum/element/processable/Attach(datum/target, tool_behaviour, result_atom_type, amount_created = 3, time_to_process = 2 SECONDS, table_required = FALSE)
+/datum/element/processable/Attach(datum/target, tool_behaviour, result_atom_type,
+	amount_created = 3, time_to_process = 2 SECONDS, table_required = FALSE, screentip_verb = "Process")
```
```suggestion
/datum/element/processable/Attach(datum/target, tool_behaviour, result_atom_type, amount_created = 3, time_to_process = 2 SECONDS, table_required = FALSE, screentip_verb = "Process")
```

it just looks odd otherwise imo

flowercuco committed:
requested format changes
adds comments to new signal that ive lached onto the add element proccessable

san7890 commented on code/datums/elements/food/processable.dm:
```diff
@@ -54,3 +61,24 @@
+ * * held_item - refers to tool used by add element proccessable to process one item into another item
+ * * user - refers to user who will see the screentip when the proper context and tool are there
+ */
+/datum/element/processable/proc/on_requesting_context_from_item(datum/source, list/context, obj/item/held_item, mob/user,)
```
```suggestion
/datum/element/processable/proc/on_requesting_context_from_item(datum/source, list/context, obj/item/held_item, mob/user)
```

Cruft when you flattened it into one line. won't affect compile but it looks bad

flowercuco committed:
Update code/datums/elements/food/processable.dm

Co-authored-by: san7890 <the@san7890.com>

flowercuco committed:
fixes 515 merge conflics

Mothblocks commented on code/modules/cargo/packs/organic.dm:
```diff
@@ -185,52 +181,58 @@
+	new_pizza_box.pizza = new randomize_pizza
+	new_pizza_box.boxtag = new_pizza_box.pizza.boxtag
+	new_pizza_box.boxtag_set = TRUE
+	new_pizza_box.update_appearance()
```
What actually needs to update here?

flowercuco replied:
the sprite doesnt start with the little text on it, so after it has a pizza in it and it knows what to write on the box it needs to update into having writing on the box

flowercuco replied:
i changed it to update the icon and the description because otherwise it has the sprite with writing but doesnt tell you what the writing is.

Mothblocks commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -117,35 +117,38 @@
 
+/obj/item/pizzabox/attack_self_secondary(mob/user)
+	if(length(boxes) > 0)
+		return
```
```suggestion
		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
```

Mothblocks commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -117,35 +117,38 @@
+	if(length(boxes) > 0)
+		return
+	if(pizza || bomb)
+		return
```
```suggestion
		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
```

Mothblocks commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -117,35 +117,38 @@
+	var/obj/item/stack/sheet/cardboard/cardboard = new /obj/item/stack/sheet/cardboard(user.drop_location())
+	balloon_alert(user, "deconstructed")
+	user.put_in_active_hand(cardboard)
+	qdel(src)
```
```suggestion
	qdel(src)
	return SECONDARY_ATTACK_CONTINUE_CHAIN
```

Mothblocks commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -117,35 +117,38 @@
+		return
+	if(pizza || bomb)
+		return
+	var/obj/item/stack/sheet/cardboard/cardboard = new /obj/item/stack/sheet/cardboard(user.drop_location())
```
```suggestion
	var/obj/item/stack/sheet/cardboard/cardboard = new(user.drop_location())
```

Mothblocks commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -117,35 +117,38 @@
+	if(pizza || bomb)
+		return
+	var/obj/item/stack/sheet/cardboard/cardboard = new /obj/item/stack/sheet/cardboard(user.drop_location())
+	balloon_alert(user, "deconstructed")
```
```suggestion
```

This isn't necessary because there is already visual feedback where the user is looking (and the user is not where they are looking anyway)

Mothblocks commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -178,28 +180,26 @@
-					to_chat(user, span_danger("You can't keep holding the stack!"))
+			if(length(boxes) >= 5)
+				if(prob(10 * length(boxes)))
+					balloon_alert(user, "dropped it!")
```
```suggestion
					balloon_alert(user, "oops!")
```

nit, i just think this reads weird

flowercuco replied:
i had no idea what to put in this one and "oops" was the simplest and best option the whole time 😔

Mothblocks commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -389,3 +389,51 @@
+/obj/item/pizzabox/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	if(!held_item)
+		if(user.get_inactive_held_item() != src)
+			return
```
```suggestion
			return NONE
```

Means the same but explicitness is good, here and others

Mothblocks commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -389,3 +389,51 @@
+			return
+		if(open)
+			if(pizza)
+				context [SCREENTIP_CONTEXT_LMB] = "Remove pizza"
```
Remove the spaces here and everywhere else in between context and the [].

flowercuco commented on code/modules/cargo/packs/organic.dm:
```diff
@@ -185,52 +181,58 @@
+	. = ..()
+	var/list/rng_pizza_list = pizza_types.Copy()
+	for(var/i in 1 to 5)
+		if(add_anomalous(new_crate))
+			continue
+		if(add_boombox(new_crate))
+			continue
```
yeah this is how it worked before, it was just less clear about it

flowercuco committed:
various formatting and syntax suggestions
changed the pizza crate update appearance into only updating icon and description
made infinite pizza box unfoldable
used pick n take now that SOMEONE told me it existed JACQUEREL

JohnFulpWillard commented on code/modules/cargo/packs/organic.dm:
```diff
@@ -155,25 +155,21 @@
-	var/anna_molly_box_chance = 1
-	///Total tickets in our figurative lottery (per pizza) to decide if we create a bomb box, and if so what type. 1 to 3 create a bomb. The rest do nothing.
-	var/boombox_tickets = 100
+	var/anomalous_box_provided = FALSE
```
Why is this not static anymore?

flowercuco replied:
it doesnt need to be, the supply console only makes the datum once

JohnFulpWillard commented on code/modules/cargo/packs/organic.dm:
```diff
@@ -185,52 +181,58 @@
+	new_pizza_box.boxtag_set = TRUE
+	new_pizza_box.update_appearance(UPDATE_ICON | UPDATE_DESC)
 
+/// tells crew that an infinite pizza box exists, half of the time, based ona roll in the anamolous box proc
```
```suggestion
/// tells crew that an infinite pizza box exists, half of the time, based on a roll in the anamolous box proc.
```

JohnFulpWillard commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -117,35 +119,38 @@
 			if(wires.is_all_cut() && bomb_defused)
 				user.put_in_hands(bomb)
-				to_chat(user, span_notice("You carefully remove the [bomb] from [src]."))
+				balloon_alert(user, "removed bomb")
```
this would balloon on the box, so if it's in their hand, it would balloon in the icon in their UI.
Though it might not be a problem since that's likely where the user would be looking, but if that's not what you want, you'll have to change it to ``user.balloon_alert``

flowercuco replied:
ive just tested it again, the balloon alert just comes out of the guy

![image](https://user-images.githubusercontent.com/116288367/202826773-c421201d-f010-40c4-8a38-35cbf38081f6.png)

Mothblocks replied:
It uses the turf of whatever it's on, it only shows in hand if it's not equipped...I think? I forget when exactly it does but bla

JohnFulpWillard commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -208,28 +210,28 @@
 			wires.interact(user)
 	else if(istype(I, /obj/item/reagent_containers/cup))
-		to_chat(user, span_warning("That's not a pizza!"))
+		balloon_alert(user, "not a pizza!")
```
I think this code is here because pizza used to be a subtype of reagent containers. You should remove it honestly.

JohnFulpWillard commented on code/modules/food_and_drinks/pizzabox.dm:
```diff
@@ -178,28 +182,26 @@
-					to_chat(user, span_danger("You can't keep holding the stack!"))
+			if(length(boxes) >= 5)
+				if(prob(10 * length(boxes)))
+					balloon_alert(user, "oops!")
```
This is a weird balloon alert, I think it would work better as either a message to chat still, or a balloon alert to all.

flowercuco replied:
changed it to a to viewers

flowercuco commented on code/datums/elements/food/processable.dm:
```diff
@@ -1,4 +1,4 @@
-// If an item has the processable item, it can be processed into another item with a specific tool. This adds generic behavior for those actions to make it easier to set-up generically.
+// If an item has the processable item, it can be processed into another item with a specific tool. This adds generic behaviour for those actions to make it easier to set-up generically.
```
there isnt a standard on english v american spelling so im going with the brits because thats how it it is called by every other variable on this document

flowercuco committed:
removed code that didnt do anything anymore apparently
turned ballon alert to a viewers alert
spelling error
